### PR TITLE
Refactor and test authn resources

### PR DESCRIPTION
**Pull request by savonarola**
Here we test auth providers with real backends.

We:
* Tidy `emqx_resource` up a bit, removing unused/not working code and adding tests.
* Try to rethink (`emqx_authenticator` <-> `emqx_authenticator` providers) interaction in the part concerning resource allocation.
* Implement `emqx_authn_redis` tests with real backends spawned with Docker Compose.

Other `emqx_authn_*` provider tests are to follow in separate PRs.

### CODE REVIEW

File: .ci/docker-compose-file/Makefile.local [hunk 1]
```diff
@@ -0,0 +1,49 @@
+.PHONY: help up down ct ct-all bash run
```
Comment (savonarola) [inline .ci/docker-compose-file/Makefile.local:1]: This is a helper `Makefile` to run tests in Docker Compose reusing CI Docker Compose configs. A user can have a full and easy to create test environment with all the resources.

We may opt not to keep it in this repo, though.

File: apps/emqx/src/emqx_authentication.erl [hunk 2]
```diff
@@ -27,6 +27,8 @@
 
 -include_lib("stdlib/include/ms_transform.hrl").
 
+-elvis([{elvis_style, god_modules, #{limit => 35}}]).
```
Comment (savonarola) [inline apps/emqx/src/emqx_authentication.erl:30]: This module requires further splitting (config handling part was moved to a new module recently), but I think this should be done in a separate PR.

File: apps/emqx/src/emqx_authentication.erl [hunk 3]
```diff
@@ -617,18 +629,17 @@ maybe_unhook(#{hooked := true} = State) ->
...
 
-do_create_authenticator(ChainName, AuthenticatorID, #{enable := Enable} = Config, Providers) ->
+do_create_authenticator(AuthenticatorID, #{enable := Enable} = Config, Providers) ->
     case maps:get(authn_type(Config), Providers, undefined) of
         undefined ->
             {error, no_available_provider};
         Provider ->
-            Unique = unique(ChainName, AuthenticatorID, ?VER_1),
-            case Provider:create(Config#{'_unique' => Unique}) of
+            case Provider:create(Config) of
```
Comment (savonarola) [inline apps/emqx/src/emqx_authentication.erl:637]: As far as I can understand, `version` is used only to generate resource ids, so that we can:
* create new resource with new id while the old is still active;
* avoid resource leakage.

I moved `resource_id` handling to the providers and added cleanup to the authentication initialization process.

File: apps/emqx/test/emqx_authentication_SUITE.erl [hunk 4]
```diff
@@ -128,23 +128,43 @@ t_authenticator(Config) when is_list(Config) ->
                              enable => true},
```
Comment (savonarola) [inline apps/emqx/test/emqx_authentication_SUITE.erl:128]: This file contains no important changes, but mostly long line splitting.

File: apps/emqx_authn/src/emqx_authn_app.erl [hunk 5]
```diff
@@ -34,6 +34,7 @@
 start(_StartType, _StartArgs) ->
     ok = mria_rlog:wait_for_shards([?AUTH_SHARD], infinity),
     {ok, Sup} = emqx_authn_sup:start_link(),
+    ok = emqx_authn_utils:cleanup_resources(),
```
Comment (savonarola) [inline apps/emqx_authn/src/emqx_authn_app.erl:37]: This cleanup doesn't make any difference now, but I added it so that we preserve it when fix `emqx_authentication` reinitialization.

File: apps/emqx_authn/src/simple_authn/emqx_authn_mnesia.erl [hunk 6]
```diff
@@ -91,6 +90,7 @@ fields(config) ->
     [ {mechanism,               {enum, ['password-based']}}
     , {backend,                 {enum, ['built-in-database']}}
     , {user_id_type,            fun user_id_type/1}
+    , {user_group,              fun user_group/1}
```
Comment (savonarola) [inline apps/emqx_authn/src/simple_authn/emqx_authn_mnesia.erl:93]: Here is an important change concerning `user_group`.

I moved it to the config part because it seems to be related to the persistent storage of users.
It seems unlikely that a user would like
* To have inevitably different users for each of his chains.
* To have all of their users _destroyed_ when they remove _authenticator config_, e.g. through API.

More likely, a user would like to have their users persistent, like with other DB providers, when destroying config destroys only DB connection.

So, `user_group` moved to config now serves more like a _storage address_, like `host:port/database`.

I may be wrong with this :)

File: apps/emqx_resource/src/emqx_resource_transform.erl [hunk 7]
```diff
@@ -1,70 +0,0 @@
...
-            {[Form], fix_spec_attrs(), fix_spec_funcs(Mod)};
-        _ ->
-            {[Form], [], []}
-    end.
-
-fix_spec_attrs() ->
-    [ ?Q("-export([emqx_resource_schema/0]).")
-    ].
-fix_spec_funcs(_Mod) ->
-    [ ?Q("emqx_resource_schema() -> <<\"demo_swagger_schema\">>.")
```
Comment (savonarola) [inline apps/emqx_resource/src/emqx_resource_transform.erl:69]: The whole `parse_transform` code served for injecting this function when including `emqx_resource/include/emqx_resource_behaviour.hrl`.

But as can be seen, the implementation of the injected fun is broken, and code relying on this fun is not working. It currently serves as a mark for a module to be a `emqx_resource` implementation.

So I resorted to the usual behaviours and their discovery.

File: apps/emqx_resource/src/emqx_resource.erl [hunk 8]
```diff
@@ -19,17 +19,13 @@
 -include("emqx_resource.hrl").
 -include("emqx_resource_utils.hrl").
 
+-elvis([{elvis_style, god_modules, #{limit => 33}}]).
```
Comment (savonarola) [inline apps/emqx_resource/src/emqx_resource.erl:22]: This module also requires splitting, but I think this should be done in a separate PR.

File: .github/workflows/run_test_cases.yaml [hunk 9]
```diff
@@ -78,18 +78,22 @@ jobs:
         - name: docker compose up
           if: env.EDITION == 'opensource'
           env:
+		        REDIS_TAG: 6
```
[the remark this answers is not in the document]
Comment (savonarola) [reply .github/workflows/run_test_cases.yaml]: ✅ (already fixed)

File: apps/emqx_authn/src/simple_authn/emqx_authn_mnesia.erl (hunk 6 above)
Comment (thalesmg) [reply apps/emqx_authn/src/simple_authn/emqx_authn_mnesia.erl:93]: Sounds like a reasonable argument to me. :slightly_smiling_face:

File: apps/emqx_authn/test/emqx_authn_redis_SUITE.erl [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+%%------------------------------------------------------------------------------
+
+t_create(_Config) ->
+    AuthConfig = raw_redis_auth_config(),
+
+    {ok, _} = emqx:update_config(
+                ?PATH,
+                {create_authenticator, ?GLOBAL, AuthConfig}),
+
+    {ok, [#{provider := emqx_authn_redis}]} = emqx_authentication:list_authenticators(?GLOBAL).
```
Comment (thalesmg) [inline apps/emqx_authn/test/emqx_authn_redis_SUITE.erl:89]: Tiny nit: maybe assert at the start of the test that this authenticator does not exist prior to the `update_config` call?

Comment (savonarola) [reply apps/emqx_authn/test/emqx_authn_redis_SUITE.erl:90]: ✅

File: apps/emqx_resource/test/emqx_test_resource.erl [hunk 11]
```diff
@@ -0,0 +1,49 @@
+-module(emqx_test_resource).
```
[the remark this answers is not in the document]
Comment (savonarola) [reply apps/emqx_resource/test/emqx_test_resource.erl:17]: ✅

File: apps/emqx/src/emqx_hooks.erl [hunk 12]
```diff
@@ -211,7 +213,7 @@ safe_execute({M, F, A}, Args) ->
                 exception => Error,
                 reason => Reason,
                 stacktrace => Stacktrace,
-                failed_call => {M, F, A}
+                failed_call => {M, F, A ++ Args}
```
[the remark this answers is not in the document]
Comment (savonarola) [reply apps/emqx/src/emqx_hooks.erl]: ✅

File: apps/emqx_resource/src/emqx_resource_api.erl [hunk 13]
```diff
@@ -15,17 +15,7 @@
...
-        , stringnify/1
-        ]).
-
-list_instances(Filter) ->
-    [format_data(Data) || Data <- emqx_resource:list_instances_verbose(), Filter(Data)].
-
-format_data(#{id := Id, mod := Mod, status := Status, config := Config}) ->
-    #{id => Id, status => Status, resource_type => Mod,
-      config => emqx_resource:call_jsonify(Mod, Config)}.
+-export([stringnify/1]).
```
[the remark this answers is not in the document]
Comment (savonarola) [reply apps/emqx_resource/src/emqx_resource_api.erl]: ✅

File: apps/emqx/src/emqx_authentication.erl [hunk 14]
```diff
@@ -341,11 +317,13 @@ create_authenticator(ChainName, Config) ->
 delete_authenticator(ChainName, AuthenticatorID) ->
     call({delete_authenticator, ChainName, AuthenticatorID}).
 
--spec update_authenticator(chain_name(), authenticator_id(), config()) -> {ok, authenticator()} | {error, term()}.
+-spec update_authenticator(chain_name(), authenticator_id(), config())
```
[the remark this answers is not in the document]
Comment (savonarola) [reply apps/emqx/src/emqx_authentication.erl]: ✅

File: apps/emqx/src/emqx_authentication.erl (hunk 2 above)
[the remark this answers is not in the document]
Comment (savonarola) [reply apps/emqx/src/emqx_authentication.erl]: ✅ moved to `elvis.config`

File: apps/emqx_resource/src/emqx_resource_transform.erl (hunk 7 above)
Comment (terry-xiaoyu) [reply apps/emqx_resource/src/emqx_resource_transform.erl:69]: This parse_transform is about to be removed. It was for generation some metadata for the supported configuration items of a resource, but the design has been changed, we use hocon schema for that now.

**Review by terry-xiaoyu — APPROVED** [review]

File: apps/emqx_authn/src/simple_authn/emqx_authn_mnesia.erl (hunk 6 above)
Comment (tigercl) [reply apps/emqx_authn/src/simple_authn/emqx_authn_mnesia.erl:93]: @savonarola Hi. In my imagination, when users create multiple authentication instances that use the built-in database, they don’t need to care about how to isolate data between multiple instances.

Now `user_group` has become a field that users must understand its usefulness, which adds complexity to a certain extent.

At the same time, we seem to lack the ability to allow users to manage these data, and how we can ensure that users will clean up these data in a timely manner.

I think the latter will be the key point.

Comment (savonarola) [reply apps/emqx_authn/src/simple_authn/emqx_authn_mnesia.erl:93]: bq. At the same time, we seem to lack the ability to allow users to manage these data, and how we can ensure that users will clean up these data in a timely manner.

Considering the introduced change, I supposed that users would be able to manage users explicitly: adding and deleting users through API.

At the same time, they wouldn't have their users (unexpectedly?) removed when they, for example, try to use an alternative backend and temporarily (as they may think) drop mnesia authentication section with API.

Comment (savonarola) [reply apps/emqx_authn/src/simple_authn/emqx_authn_mnesia.erl:93]: @tigercl if we want to isolate users by default, we may use authenticator id as default `group_id` (like it was previously done).

**Review by zmstone — DISMISSED** [review]

File: apps/emqx_authn/src/enhanced_authn/emqx_enhanced_authn_scram_mnesia.erl [hunk 15]
```diff
@@ -131,14 +135,7 @@ authenticate(#{auth_method := AuthMethod,
...
 
-destroy(#{user_group := UserGroup}) ->
-    trans(
-        fun() ->
-            MatchSpec = [{{user_info, {UserGroup, '_'}, '_', '_', '_', '_'}, [], ['$_']}],
-            ok = lists:foreach(fun(UserInfo) ->
-                                  mnesia:delete_object(?TAB, UserInfo, write)
-                               end, mnesia:select(?TAB, MatchSpec, write))
-        end).
+destroy(_St) -> ok.
```
Comment (zmstone) [inline apps/emqx_authn/src/enhanced_authn/emqx_enhanced_authn_scram_mnesia.erl:138]: Q: intended to ignore this callback ?

**Review by zmstone — APPROVED** [review]

File: apps/emqx_authn/src/enhanced_authn/emqx_enhanced_authn_scram_mnesia.erl (hunk 15 above)
Comment (savonarola) [reply apps/emqx_authn/src/enhanced_authn/emqx_enhanced_authn_scram_mnesia.erl:138]: No, we are return it back too.
I should have drafted PR not to bother with updates before final review 😓

**Review by zmstone — APPROVED** [review]
